File: CSharpCodeBase/aiactions/wander.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class wander :ainode{
 var AI_Wander = Class({__includes=AINode, init = public void (self, entity){
   AINode.init(self, entity);
   this.wanderPosition = null ;
   this.timeout = null;
   this.wanderRadius = this.entity.config:Get("wanderRadius");
 }})
 public void AI_Visit(){
   if(this.status == NODE_READY  ){ 
     // Choose new position
     self:NewPosition();
     this.status = NODE_RUNNING;
     this.entity.mover:ResetSpeed();
     this.entity.mover:SetGoal(this.wanderPosition);
   }else{if this.status == NODE_RUNNING  ){ 
     if(not this.entity.mover:IsHaveGoal()  ){ 
       if(this.idleTimeout > 0  ){ 
         this.idleTimeout = this.idleTimeout - GameController.deltaTime;
         if(this.idleTimeout <= 0  ){ 
           this.entity.mecanim:SetFloat("idle_type", math.random(0, this.entity.config:Get("wanderIdleTypes")));
           this.entity.mecanim:SetTrigger("idle");
         }
       }
       if(this.timeout > 0  ){ 
         this.timeout = this.timeout - GameController.deltaTime;
       }else{
         this.status = NODE_SUCCESS;
       }
     }
   }
   return this.status;
 }
 public void AI_NewPosition(){
   var aroundPosition = this.entity.worldmap:GetLocation("home");
   if(this.entity.isChild  &&  this.entity.relationship:GetInstance("parent") != null  ){ 
     aroundPosition = this.entity.relationship:GetInstance("parent"):GetPosition();
   }
   this.wanderPosition = aroundPosition;
     + Vector3(math.random(-this.wanderRadius, this.wanderRadius), ;
         0,;
         math.random(-this.wanderRadius, this.wanderRadius));
   this.timeout = math.random(this.entity.config:Get("wanderIdleMin"),;
     this.entity.config:Get("wanderIdleMax"));
   this.idleTimeout = math.random(0, this.timeout);
 }
 return AI_Wander}}

File: CSharpCodeBase/aiactions/wander.lua
local Class = require 'utils.hump.class'
local Vector3 = require 'utils.hump.vector3'
local AINode = require 'utils.ailib.ainode'

local AI_Wander = Class({__includes=AINode, init = function (self, entity)
  AINode.init(self, entity)
  self.wanderPosition = nil 
  self.timeout = nil
  self.wanderRadius = self.entity.config:Get("wanderRadius")
end})

function AI_Wander:Visit()
  if self.status == NODE_READY then
    -- Choose new position
    self:NewPosition()
    self.status = NODE_RUNNING
    self.entity.mover:ResetSpeed()
    self.entity.mover:SetGoal(self.wanderPosition)
  elseif self.status == NODE_RUNNING then
    if not self.entity.mover:IsHaveGoal() then
      if self.idleTimeout > 0 then
        self.idleTimeout = self.idleTimeout - GameController.deltaTime
        if self.idleTimeout <= 0 then
          self.entity.mecanim:SetFloat('idle_type', math.random(0, self.entity.config:Get('wanderIdleTypes')))
          self.entity.mecanim:SetTrigger('idle')
        end
      end
      if self.timeout > 0 then
        self.timeout = self.timeout - GameController.deltaTime
      else
        self.status = NODE_SUCCESS
      end
    end
  end
  return self.status
end

function AI_Wander:NewPosition()
  local aroundPosition = self.entity.worldmap:GetLocation('home')
  if self.entity.isChild and self.entity.relationship:GetInstance('parent') ~= nil then
    aroundPosition = self.entity.relationship:GetInstance('parent'):GetPosition()
  end
  self.wanderPosition = aroundPosition
    + Vector3(math.random(-self.wanderRadius, self.wanderRadius), 
        0,
        math.random(-self.wanderRadius, self.wanderRadius))
  self.timeout = math.random(self.entity.config:Get("wanderIdleMin"),
    self.entity.config:Get("wanderIdleMax"))
  self.idleTimeout = math.random(0, self.timeout)
end

return AI_Wander